### ISSUE TITLE
Calculate Lagrange points

### ISSUE THREAD

**Issue by cosinekitty** [cosinekitty/astronomy#148]
It might be interesting to calculate Lagrange points for each planet with respect to the Sun.

**Comment by gerben-b** [comment]
I was also wondering if Earth's L2 and james webb could be calculated. I did some searching but hand worked out the math yet.

**Comment by cosinekitty** [comment]
@gerben-b When I read about Lagrange points, it looks like there is a simple geometric pattern. Three of them are on a line connecting the Sun and Earth, and the other two are +60 degrees and -60 degrees around the orbital plane from the Earth.

The good news I can generate test data using the [JPL Horizons](https://ssd.jpl.nasa.gov/horizons/app.html#/) tool. I asked the tool to generate state vectors (position and velocity) for Sun-Earth-Moon L2 using their object called "SEMB-L2". 

Here is how they describe SEMB-L2:

```
Revised: Aug 29, 2013                SEMB-L2                                32
                      (Sun & Earth-Moon Barycenter Lagrange 2)

  #1) The Sun & Earth-Moon Barycenter Lagrange-2 (L2) point is a location 
       where the Earth's gravitational field partially counters that of the 
       Sun.

  #2) This L2 point is about 1.5 million km (~900,000 miles) away from the 
       Earth, opposite the direction of the Sun, or slightly less than one 
       percent of the Earth-Sun distance (four times the distance from Earth 
       to the Moon) 

  #3) L2 has been selected as the location of the next generation James Webb 
       Space Telescope, was used by the Genesis spacecraft on the return to 
       Earth, and is (or will be) used for WMAP, Herschel, and Planck 
       spacecraft. 
```

Since I can generate reference position and velocity vectors for Lagrange points, that means I have a way to check any code I write to calculate them, to see how close I am to the correct answers.

**Comment by gerben-b** [comment]
I think you're right. Calculating the Lagrange points should be relatively straight forward. JWST, however, is orbiting L2 in an orbit roughly the size of the Moon's orbit around Earth. This is, of course, a bit off topic for this issue, but I was just curious :-).

**Comment by cosinekitty** [comment]
Here is a good video that was sent to me regarding the James Webb orbit: 

https://www.youtube.com/watch?v=ybn8-_QV8Tg

Around 8:55 into the video it explains that the spacecraft does not exactly orbit L2, but a point slightly closer to the Earth than L2, and explains why that is necessary. Interesting stuff!

**Comment by gerben-b** [comment]
Thats interesting and he explains it very well, thank you.

**Comment by cosinekitty** [comment]
I have been looking deeper into calculating Lagrange points. Here are some initial notes.

- I found guidance on the math here:
https://www.mat.univie.ac.at/~westra/lagrangepoints.pdf
It is an interesting read!

- When I use JPL Horizons to generate test data, I notice they support L1, L2, L4, and L5, but not L3. In the Sun/Earth case, I suppose L3 is useless for spacecraft, because the spacecraft would be hidden by the Sun and thus impossible to communicate with from Earth. (Maybe L3 is where the lizard people hide their mother ship from us?) I have a policy of not adding a feature to Astronomy Engine unless I can find reliable test data and I write an automated test based on that data. However, I may make an exception for L3 if tests for the other four points can be confirmed, because the L3 point looks trivial to calculate.

- JPL Horizons provides Sun/Earth and Earth/Moon Lagrange points, but for not any other body pairs (as far as I can tell). However, it would be interesting to support other planets with respect to the Sun, especially Jupiter. So this is another case where I may bend the rules; if I generalize the formulas to work for Sun/Earth and Earth/Moon, and those tests pass, then perhaps we can trust they also work for other bodies. It would be the same code, just using different position vectors and body masses fed into them.

Any comments or feedback from the community would be welcome, before I go much further.

**Comment by gerben-b** [comment]
Don't worry about the lizard people, I went there to check and they assured me they have no inappropriate intentions with us. Dennis Westra's paper is interesting indeed. I am not very good at mathematics but I think I'll give this a try myself, just to see how far I can get.

**Comment by cosinekitty** [comment]
I'm very close to having a C version of a calculator for Lagrange point state vectors. Some notes that might be of interest:

1. The Westra paper is helpful for finding approximate solutions for L1..L3, but not exact solutions. I ended up using his formulas to find the initial guess for the radial distance between the bodies. Then I iterate Newton's method to find an exact solution. It generally converges in fewer than 10 iterations. The formula models a rotating frame of reference that moves with the same angular speed as the two bodies co-orbiting their common barycenter. It maps each position along the line passing through both bodies' centers to the total acceleration there: two gravitational accelerations and one centrifugal acceleration. L1, L2, L3 are the three locations along this rotating axis where the accelerations add up to zero.
2. Almost all of my calculations match JPL Horizons calculations, but not all. I agree with JPL Horizons for all Lagrange point position vectors: L1..L5. I also agree with JPL Horizons L1..L3 velocity vectors. I agree with JPL Horizons L4/L5 velocity vectors for the Sun/EMB system. The only discrepancy is with L4/L5 velocity vectors for the Earth/Moon system, in which case we disagree by 5.4 arcminutes of direction, but the vector magnitudes (speeds) agree. This angular oddity appears in the JPL Horizons data itself: their velocity vectors point 5.4 arcminutes outside the plane passing through the Earth/Moon/L4 equilateral triangle (and the same for Earth/Moon/L5). This seems wrong, but it's most likely something I don't understand yet. I sent an email to the maintainer of JPL Horizons about this, and I'm waiting for a response before I proceed.
